File: scripts/get-firebase-token.mjs
#!/usr/bin/env node

/**
 * Get Firebase ID Token for API Testing
 * 
 * Usage:
 *   node scripts/get-firebase-token.mjs email@example.com password123
 *   
 * Or set environment variables:
 *   FIREBASE_EMAIL=email@example.com
 *   FIREBASE_PASSWORD=password123
 *   node scripts/get-firebase-token.mjs
 */

import { initializeApp } from 'firebase/app';
import { getAuth, signInWithEmailAndPassword } from 'firebase/auth';
import { readFileSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

// Load Firebase config from your project
const serviceAccount = JSON.parse(
  readFileSync(join(__dirname, '../serviceAccountKey.json'), 'utf8')
);

// Extract project ID from service account
const projectId = serviceAccount.project_id;

// Firebase config (you need to add your web config here)
const firebaseConfig = {
  apiKey: process.env.FIREBASE_API_KEY || "YOUR_WEB_API_KEY",
  authDomain: `${projectId}.firebaseapp.com`,
  projectId: projectId,
};

// Get credentials from command line or environment
const email = process.argv[2] || process.env.FIREBASE_EMAIL;
const password = process.argv[3] || process.env.FIREBASE_PASSWORD;

if (!email || !password) {
  console.error('❌ Error: Email and password required');
  console.log('\nUsage:');
  console.log('  node scripts/get-firebase-token.mjs email@example.com password123');
  console.log('\nOr set environment variables:');
  console.log('  export FIREBASE_EMAIL=email@example.com');
  console.log('  export FIREBASE_PASSWORD=password123');
  console.log('  node scripts/get-firebase-token.mjs');
  process.exit(1);
}

console.log('🔐 Signing in to Firebase...');
console.log(`📧 Email: ${email}`);

// Initialize Firebase
const app = initializeApp(firebaseConfig);
const auth = getAuth(app);

try {
  // Sign in
  const userCredential = await signInWithEmailAndPassword(auth, email, password);
  const user = userCredential.user;
  
  // Get ID token
  const idToken = await user.getIdToken();
  
  console.log('\n✅ Successfully authenticated!');
  console.log('\n📋 User Info:');
  console.log(`   UID: ${user.uid}`);
  console.log(`   Email: ${user.email}`);
  console.log(`   Email Verified: ${user.emailVerified}`);
  
  console.log('\n🎫 Firebase ID Token:');
  console.log('─'.repeat(80));
  console.log(idToken);
  console.log('─'.repeat(80));
  
  console.log('\n📌 How to use in Postman:');
  console.log('1. Open your Postman collection');
  console.log('2. Go to Variables tab');
  console.log('3. Set firebase_token = [copy token above]');
  console.log('4. Or use Authorization → Bearer Token → {{firebase_token}}');
  
  console.log('\n⏰ Token expires in: 1 hour');
  console.log('   Run this script again when it expires');
  
  console.log('\n🔧 Quick test:');
  console.log(`   curl -H "Authorization: Bearer ${idToken.substring(0, 50)}..." \\`);
  console.log('        http://localhost:5055/api/me');
  
  process.exit(0);
  
} catch (error) {
  console.error('\n❌ Authentication failed:', error.message);
  
  if (error.code === 'auth/invalid-email') {
    console.log('💡 The email address is badly formatted');
  } else if (error.code === 'auth/user-not-found') {
    console.log('💡 No user found with this email');
  } else if (error.code === 'auth/wrong-password') {
    console.log('💡 Incorrect password');
  } else if (error.code === 'auth/too-many-requests') {
    console.log('💡 Too many failed login attempts. Try again later.');
  }
  
  process.exit(1);
}

File: scripts/get-firebase-token.py
#!/usr/bin/env python3

"""
Get Firebase ID Token - Simple Python Version
Usage: python3 scripts/get-firebase-token.py EMAIL PASSWORD
"""

import sys
import json
import urllib.request
import urllib.parse

# Firebase Web API Key
API_KEY = "AIzaSyDFzUfv1enm5_lucOMz4tWh26GJfIG751M"

if len(sys.argv) < 3:
    print("❌ Error: Email and password required")
    print("\nUsage:")
    print("  python3 scripts/get-firebase-token.py EMAIL PASSWORD")
    print("\nExample:")
    print('  python3 scripts/get-firebase-token.py 22onsloanedigitalteam@gmail.com "Sloane22Gen!"')
    sys.exit(1)

email = sys.argv[1]
password = sys.argv[2]

print("🔐 Signing in to Firebase...")
print(f"📧 Email: {email}")
print()

# Firebase Auth REST API
url = f"https://identitytoolkit.googleapis.com/v1/accounts:signInWithPassword?key={API_KEY}"
data = {
    "email": email,
    "password": password,
    "returnSecureToken": True
}

try:
    # Make request
    req = urllib.request.Request(
        url,
        data=json.dumps(data).encode('utf-8'),
        headers={'Content-Type': 'application/json'}
    )
    
    with urllib.request.urlopen(req) as response:
        result = json.loads(response.read().decode('utf-8'))
    
    # Success!
    print("✅ Successfully authenticated!")
    print()
    print("📋 User Info:")
    print(f"   Email: {result['email']}")
    print(f"   UID: {result['localId']}")
    print(f"   Expires in: {result['expiresIn']}s (1 hour)")
    print()
    print("🎫 Firebase ID Token:")
    print("─" * 80)
    print(result['idToken'])
    print("─" * 80)
    print()
    print("🔄 Refresh Token (save for later):")
    print(result['refreshToken'][:50] + "...")
    print()
    print("📌 How to use in Postman:")
    print("1. Open your collection → Variables tab")
    print("2. Set firebase_token = [copy token above]")
    print("3. Test with any authenticated endpoint")
    print()
    print("🔧 Quick test command:")
    token_preview = result['idToken'][:50]
    print(f'   curl -H "Authorization: Bearer {token_preview}..." \\')
    print("        http://localhost:5055/api/me")
    print()
    
    # Save to file option
    save = input("💾 Save token to .env.firebase-token file? [y/N]: ").strip().lower()
    if save == 'y':
        from datetime import datetime
        with open('.env.firebase-token', 'w') as f:
            f.write(f"# Firebase Authentication Token\n")
            f.write(f"# Generated: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}\n")
            f.write(f"# Expires in: 1 hour\n\n")
            f.write(f"FIREBASE_TOKEN={result['idToken']}\n")
            f.write(f"FIREBASE_REFRESH_TOKEN={result['refreshToken']}\n")
            f.write(f"FIREBASE_USER_ID={result['localId']}\n")
            f.write(f"FIREBASE_EMAIL={result['email']}\n")
        print("✅ Saved to .env.firebase-token")
        print("   Source it with: source .env.firebase-token")

except urllib.error.HTTPError as e:
    error_data = json.loads(e.read().decode('utf-8'))
    print("❌ Authentication failed:")
    print(json.dumps(error_data, indent=2))
    print()
    
    if 'error' in error_data:
        error_msg = error_data['error'].get('message', '')
        
        if 'INVALID_LOGIN_CREDENTIALS' in error_msg or 'INVALID_PASSWORD' in error_msg:
            print("💡 Wrong email or password")
            print("   Check your credentials or reset password")
        elif 'EMAIL_NOT_FOUND' in error_msg:
            print("💡 No user found with this email")
            print("   Create user in Firebase Console → Authentication")
        elif 'USER_DISABLED' in error_msg:
            print("💡 This user account is disabled")
        elif 'TOO_MANY_ATTEMPTS' in error_msg:
            print("💡 Too many failed attempts")
            print("   Try again later")
    
    sys.exit(1)

except Exception as e:
    print(f"❌ Error: {str(e)}")
    sys.exit(1)
